torch_xla/distributed/spmd/debugging.py

```diff
@@ -0,0 +1,207 @@
+from collections.abc import Sequence
+import functools
+import string
+import sys
+from typing import Any, Callable, Optional, Union
+import weakref
+
+import numpy as np
+import torch
+import torch_xla.core.xla_model as xm
+import torch_xla.runtime as xr
+import torch_xla.experimental.xla_sharding as xs
+from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
+
+try:
+  import rich
+  import rich.align
+  import rich.box
+  import rich.console
+  import rich.padding
+  import rich.style
+  import rich.table
+  RICH_ENABLED = True
+except:
+  RICH_ENABLED = False
+
+# Sharding visualization
+sharding_callbacks = weakref.WeakValueDictionary()
+_INSPECT_SHARDING_CALL_NAME = "InspectSharding"
+
+
+class ShardingCallbackInfo:
+
+  def __init__(self, callback, module_context):
+    self.callback = callback
+    self.module_context = module_context
+
+
+Color = Union[tuple[float, float, float], str]
+ColorMap = Callable[[float], tuple[float, float, float, float]]
+
+
+def _canonicalize_color(color: Color) -> str:
+  if isinstance(color, str):
+    return color
+  r, g, b = (int(a * 255) for a in color)
+  return f"#{r:02X}{g:02X}{b:02X}"
+
+
+def _get_text_color(color: str) -> str:
+  r, g, b = torch.map(lambda x: int(x, 16),
+                      (color[1:3], color[3:5], color[5:7]))
+  if (r * 0.299 + g * 0.587 + b * 0.114) > 186:
+    return "#000000"
+  return "#ffffff"
+
+
+def make_color_iter(color_map, num_rows, num_cols):
+  num_colors = num_rows * num_cols
+  color_values = np.linspace(0, 1, num_colors)
+  idx = 0
+  for _ in range(num_colors):
+    yield color_map(color_values[idx])
+    idx = (idx + num_colors // 2 + bool(num_colors % 2 == 0)) % num_colors
+
+
+def visualize_sharding(shape: torch.Size,
```

Comment:
Why do we need tensor shape here? Does this mean we support
- 2D tensor over 2D mesh

but, not support
- 3D tensor over 2D mesh
?

We are visualizing the shard placements, and the shape/content of each shard is not shown. We just work with `sharding`(str) like 
```
# sharding is longer than 0
    # eg: '{devices=[2,2]0,1,2,3}' # 13
    # eg: '{replicated}'
    # eg: '{devices=[2,1,2]0,1,2,3 last_tile_dim_replicate}' # 15
    ```

Reply:
thanks, updated to visualize based on sharding info only, deleted the shape parameter here